grakn-graql/src/main/java/ai/grakn/graql/internal/query/QueryAnswer.java

```diff
@@ -156,11 +155,40 @@ public Answer merge(Answer a2, boolean mergeExplanation){
         if(a2.isEmpty()) return this;
         if(this.isEmpty()) return a2;
 
+        Sets.SetView<Var> varUnion = Sets.union(this.vars(), a2.vars());
+        Set<Var> varIntersection = Sets.intersection(this.vars(), a2.vars());
+        Map<Var, Concept> entryMap = Sets.union(
+                this.entrySet(),
+                a2.entrySet()
+        )
+                .stream()
+                .filter(e -> !varIntersection.contains(e.getKey()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+        varIntersection
+                .forEach(var -> {
+                    Concept concept = this.get(var);
+                    Concept otherConcept = a2.get(var);
+                    if (concept.equals(otherConcept)) entryMap.put(var, concept);
```

Comment:
Maybe we could put this in another method? The indentation and the embedding in the lambda makes it very hard to understand. I had to read this one a few times.